provider/boltmem/boltmem.go

```diff
@@ -0,0 +1,426 @@
+
+		for k, v := c.First(); k != nil; k, v = c.Next() {
+			var a types.Alert
+			if err := json.Unmarshal(v, &a); err != nil {
+				return err
+			}
+			alerts = append(alerts, &a)
+		}
+
+		return nil
+	})
+	return alerts, err
+}
+
+// Get returns the alert for a given fingerprint.
+func (a *Alerts) Get(fp model.Fingerprint) (*types.Alert, error) {
+	var alert types.Alert
+	err := a.db.View(func(tx *bolt.Tx) error {
+		b := tx.Bucket(bktAlerts)
+
+		fpb := make([]byte, 8)
+		binary.BigEndian.PutUint64(fpb, uint64(fp))
+
+		ab := b.Get(fpb)
+		if ab == nil {
+			return provider.ErrNotFound
+		}
+
+		return json.Unmarshal(ab, &alert)
+	})
+	return &alert, err
+}
+
+// Put adds the given alert to the set.
+func (a *Alerts) Put(alerts ...*types.Alert) error {
+	a.mtx.Lock()
+	defer a.mtx.Unlock()
+
+	err := a.db.Update(func(tx *bolt.Tx) error {
+		b := tx.Bucket(bktAlerts)
+
+		for _, alert := range alerts {
+			fp := make([]byte, 8)
+			binary.BigEndian.PutUint64(fp, uint64(alert.Fingerprint()))
+
+			ab := b.Get(fp)
+
+			// Merge the alert with the existing one.
+			if ab != nil {
+				var old types.Alert
+				if err := json.Unmarshal(ab, &old); err != nil {
+					return fmt.Errorf("decoding alert failed: %s", err)
+				}
+				// Merge alerts if there is an overlap in activity range.
+				if (alert.EndsAt.Sub(old.StartsAt) > 0 && alert.EndsAt.Sub(old.EndsAt) < 0) ||
+					(alert.StartsAt.Sub(old.StartsAt) > 0 && alert.StartsAt.Sub(old.EndsAt) < 0) {
+					alert = old.Merge(alert)
+				}
+			}
+
+			ab, err := json.Marshal(alert)
+			if err != nil {
+				return fmt.Errorf("encoding alert failed :%s", err)
+			}
+
+			if err := b.Put(fp, ab); err != nil {
+				return fmt.Errorf("writing alert failed: %s", err)
+			}
+
+			// Send the update to all subscribers.
+			for _, ch := range a.listeners {
+				ch <- alert
+			}
+		}
+		return nil
+	})
+
+	return err
+}
+
+// Silences gives access to silences. All methods are goroutine-safe.
+type Silences struct {
+	db *bolt.DB
+	mk types.Marker
+}
+
+func NewSilences(path string, mk types.Marker) (*Silences, error) {
+	db, err := bolt.Open(filepath.Join(path, "silences.db"), 0666, nil)
+	if err != nil {
+		return nil, err
+	}
+	err = db.Update(func(tx *bolt.Tx) error {
+		_, err := tx.CreateBucketIfNotExists(bktSilences)
+		return err
+	})
+	return &Silences{db: db, mk: mk}, err
+}
+
+func (s *Silences) Close() error {
+	return s.db.Close()
+}
+
+// The Silences provider must implement the Muter interface
+// for all its silences. The data provider may have access to an
+// optimized view of the data to perform this evaluation.
+func (s *Silences) Mutes(lset model.LabelSet) bool {
+	sils, err := s.All()
+	if err != nil {
+		log.Errorf("retrieving silences failed: %s", err)
+		// In doubt, do not silence anything.
+		return false
+	}
+
+	for _, sil := range sils {
+		if sil.Mutes(lset) {
+			s.mk.SetSilenced(lset.Fingerprint(), sil.ID)
+			return true
+		}
+	}
+
+	s.mk.SetSilenced(lset.Fingerprint())
```

Comment:
Out of scope for this PR, but I find this interface confusing. It does two different things based on whether there is a variadic arg (while the method name doesn't make that clear), but then it also only cares about the first variadic arg and ignores any further ones.

I think two separate methods would be easier to understand.